File: core_engine/core_engine_c/sandbox.c
/* 
 * KorVM: Production-Grade Zero-Trust Linear Memory Sandbox (C Core)
 * Author: Elif Nur Ayhan (codebygunes)
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdbool.h>

#define WASM_PAGE_SIZE (64 * 1024)

typedef struct {
    uint8_t *memory;
    size_t size;
} ZeroTrustSandbox;

ZeroTrustSandbox* sandbox_init(size_t initial_pages) {
    ZeroTrustSandbox *sb = (ZeroTrustSandbox*)malloc(sizeof(ZeroTrustSandbox));
    sb->size = initial_pages * WASM_PAGE_SIZE;
    sb->memory = (uint8_t*)calloc(1, sb->size);
    printf("[+] C Sandbox Initialized: %zu bytes allocated.\n", sb->size);
    return sb;
}

void sandbox_free(ZeroTrustSandbox *sb) {
    if (sb) {
        free(sb->memory);
        free(sb);
    }
}

bool check_bounds(ZeroTrustSandbox *sb, size_t offset, size_t size) {
    if (offset + size < offset) return false; // Integer overflow check
    if (offset + size > sb->size) return false;
    return true;
}

int store_i32(ZeroTrustSandbox *sb, size_t offset, int32_t value) {
    if (!check_bounds(sb, offset, 4)) {
        printf("[SUCCESS] Caught violation as expected: ZERO-TRUST VIOLATION!\n");
        return -1;
    }
    memcpy(sb->memory + offset, &value, 4);
    return 0;
}

int load_i32(ZeroTrustSandbox *sb, size_t offset, int32_t *out_val) {
    if (!check_bounds(sb, offset, 4)) {
        printf("[SUCCESS] Caught read violation as expected!\n");
        return -1;
    }
    memcpy(out_val, sb->memory + offset, 4);
    return 0;
}

int main() {
    printf("--- KorVM C Core Verification ---\n");
    ZeroTrustSandbox *sb = sandbox_init(1);

    store_i32(sb, 0, 1337);
    int32_t val = 0;
    load_i32(sb, 0, &val);
    printf("[+] Success: Stored and loaded value: %d\n", val);

    printf("\n[*] Testing illegal write at boundary...\n");
    store_i32(sb, WASM_PAGE_SIZE - 2, 42);

    sandbox_free(sb);
    return 0;
}

File: core_engine/src/sandbox.rs
/// KorVM: Production-Grade Zero-Trust Linear Memory Sandbox (Rust Core)
/// Author: Elif Nur Ayhan (codebygunes)

pub struct ZeroTrustSandbox {
    memory: Vec<u8>,
    max_pages: Option<usize>,
}

const WASM_PAGE_SIZE: usize = 64 * 1024; // 64 KB

impl ZeroTrustSandbox {
    pub fn new(initial_pages: usize, max_pages: Option<usize>) -> Self {
        let memory_size = initial_pages * WASM_PAGE_SIZE;
        println!("[+] Rust Sandbox Initialized: {} bytes allocated.", memory_size);
        Self {
            memory: vec![0; memory_size],
            max_pages,
        }
    }

    #[inline(always)]
    pub fn check_bounds(&self, offset: usize, size: usize) -> Result<(), &'static str> {
        if offset.checked_add(size).map_or(true, |end| end > self.memory.len()) {
            return Err("ZERO-TRUST VIOLATION: Illegal memory access attempt!");
        }
        Ok(())
    }

    pub fn store_i32(&mut self, offset: usize, value: i32) -> Result<(), &'static str> {
        self.check_bounds(offset, 4)?;
        let bytes = value.to_le_bytes();
        self.memory[offset..offset + 4].copy_from_slice(&bytes);
        Ok(())
    }

    pub fn load_i32(&self, offset: usize) -> Result<i32, &'static str> {
        self.check_bounds(offset, 4)?;
        let bytes: [u8; 4] = self.memory[offset..offset + 4]
            .try_into()
            .map_err(|_| "Memory conversion error")?;
        Ok(i32::from_le_bytes(bytes))
    }
}